### PR TITLE
Dat 1254 get declined invoices endpoint

### PR DESCRIPTION
**Pull request by ms-fcabrera**
Add new endpoint to get the declined invoices of a user and the total amount of debt the user has.

[DAT-1245](https://makingsense.atlassian.net/browse/DAT-1254)

### CODE REVIEW

File: Doppler.BillingUser/Controllers/BillingController.cs [hunk 1]
```diff
@@ -255,6 +255,44 @@ public async Task<IActionResult> UpdateCurrentPaymentMethod(string accountname,
...
+        {
+            var userBillingInfo = await _userRepository.GetUserBillingInformation(accountname);
+
+            if (userBillingInfo.PaymentMethod != PaymentMethodEnum.CC && userBillingInfo.PaymentMethod != PaymentMethodEnum.MP)
+            {
+                return new BadRequestObjectResult("Payment method not supported")
+                {
+                    StatusCode = 500
+                };
+            }
```
Comment (andresmoschini) [inline Doppler.BillingUser/Controllers/BillingController.cs:270]: Is it necessary? The invoice is declined independently of the payment method, right?

File: Doppler.BillingUser/Controllers/BillingController.cs [hunk 2]
```diff
@@ -255,6 +255,44 @@ public async Task<IActionResult> UpdateCurrentPaymentMethod(string accountname,
             }
         }
 
+        [Authorize(Policies.PROVISORY_USER_OR_SUPER_USER)]
+        [HttpGet("/accounts/{accountname}/declined-invoices")]
```
Comment (andresmoschini) [inline Doppler.BillingUser/Controllers/BillingController.cs:259]: What do you think about 

```suggestion
        [HttpGet("/accounts/{accountname}/invoices/declined")]
```

Or 

```suggestion
        [HttpGet("/accounts/{accountname}/invoices?declined=true")]
```

File: Doppler.BillingUser/Controllers/BillingController.cs [hunk 3]
```diff
@@ -255,6 +255,44 @@ public async Task<IActionResult> UpdateCurrentPaymentMethod(string accountname,
...
+                    InvoiceNumber = invoice.InvoiceNumber,
+                    Amount = invoice.Amount
+                });
+            }
+
+            return new OkObjectResult(new GetDeclinedInvoicesResult()
+            {
+                TotalPending = total,
+                Invoices = declinedInvoicesData
+            });
```
Comment (andresmoschini) [inline Doppler.BillingUser/Controllers/BillingController.cs:293]: In general, when it is possible, using declarative Linq code is better than imperative loop mutation a collection. 

For example:

```suggestion
            var declinedInvoicesData = invoices.Select(invoice => new DeclinedInvoiceData()
            {
                Date = invoice.Date,
                InvoiceNumber = invoice.InvoiceNumber,
                Amount = invoice.Amount
            })
            .ToList();

            return new OkObjectResult(new GetDeclinedInvoicesResult()
            {
                TotalPending = declinedInvoicesData.Sum(x => x.Amount),
                Invoices = declinedInvoicesData
            });
```

File: Doppler.BillingUser/Controllers/BillingController.cs [hunk 4]
```diff
@@ -300,7 +338,21 @@ public async Task<IActionResult> Reprocess(string accountname)
...
+
+                foreach (var invoice in invoices)
+                {
+                    failedInvoices.Add(new FailedToReprocessInvoice()
+                    {
+                        Amount = invoice.Amount,
+                        InvoiceNumber = invoice.InvoiceNumber,
+                        Error = "" // Should be invoice.ErrorMesagge?
+                    });
+                }
```
Comment (andresmoschini) [inline Doppler.BillingUser/Controllers/BillingController.cs:353]: The same here:

```suggestion
                var failedInvoices = invoices.Select(invoice => new FailedToReprocessInvoice()
                {
                    Amount = invoice.Amount,
                    InvoiceNumber = invoice.InvoiceNumber,
                    Error = "" // Should be invoice.ErrorMesagge?
                });
```

File: Doppler.BillingUser/Controllers/BillingController.cs (hunk 3 above)
Comment (ms-fcabrera) [reply Doppler.BillingUser/Controllers/BillingController.cs:293]: Done!

**Comment by jhoffman-ms** [comment]
👍

**Pull request merged**